syntax = "proto3";
package gladys.v1;

import "google/protobuf/struct.proto";
import "google/protobuf/timestamp.proto";
import "types.proto";

// ============================================================================
// REQUEST METADATA (included in all RPCs)
// ============================================================================

message RequestMetadata {
    string request_id = 1;          // UUID, generated at origin
    string trace_id = 2;            // OpenTelemetry trace ID
    string span_id = 3;             // OpenTelemetry span ID
    int64 timestamp_ms = 4;         // Request creation time (Unix ms)
    string source_component = 5;    // Originating component ID
}

// ============================================================================
// EVENTS
// ============================================================================

message Event {
    string id = 1;                              // UUID
    google.protobuf.Timestamp timestamp = 2;
    string source = 3;                          // Sensor ID

    // Content
    string raw_text = 4;                        // Natural language description
    google.protobuf.Struct structured = 5;     // Domain-specific fields

    // Salience (populated by Salience Gateway)
    gladys.types.SalienceVector salience = 6;

    // Entity references (populated by Entity Extractor)
    repeated string entity_ids = 7;

    // Optional pre-tokenized payload (for LLM injection)
    repeated int32 tokens = 8;
    string tokenizer_id = 9;

    // Heuristic that matched this event (for TD learning feedback)
    string matched_heuristic_id = 10;

    // Metadata
    RequestMetadata metadata = 15;
}

// SalienceVector moved to types.proto (gladys.types.SalienceVector)
// Single source of truth for cross-service types

// A batch of events accumulated by Orchestrator and sent to Executive on tick.
// Events are timestamp-ordered with salience already evaluated and attached.
message Moment {
    repeated Event events = 1;                  // Events with salience populated
    google.protobuf.Timestamp start_time = 2;  // First event timestamp
    google.protobuf.Timestamp end_time = 3;    // Last event timestamp (or flush time)
}

// ============================================================================
// COMPONENT STATUS
// ============================================================================

enum ComponentState {
    COMPONENT_STATE_UNKNOWN = 0;
    COMPONENT_STATE_STARTING = 1;
    COMPONENT_STATE_ACTIVE = 2;
    COMPONENT_STATE_PAUSED = 3;
    COMPONENT_STATE_STOPPING = 4;
    COMPONENT_STATE_STOPPED = 5;
    COMPONENT_STATE_ERROR = 6;
    COMPONENT_STATE_DEAD = 7;
}

message ComponentStatus {
    string component_id = 1;
    ComponentState state = 2;
    string message = 3;             // Human-readable status
    google.protobuf.Timestamp last_heartbeat = 4;
    map<string, string> metrics = 5;  // Key-value metrics
}

// ============================================================================
// ERRORS
// ============================================================================

message ErrorDetail {
    string code = 1;                // Machine-readable error code
    string message = 2;             // Human-readable message
    map<string, string> metadata = 3;
}
